errors/context.go

```diff
@@ -0,0 +1,193 @@
+package errors
+
+import (
+	"bytes"
+	"errors"
+	"fmt"
+	"io"
+
+	"github.com/mailgun/holster/v5/callstack"
+	"github.com/sirupsen/logrus"
+)
+
+// HasContext Implement this interface to pass along unstructured context to the logger
+type HasContext interface {
+	Context() map[string]interface{}
+}
+
+// HasFormat True if the interface has the format method (from fmt package)
+type HasFormat interface {
+	Format(st fmt.State, verb rune)
+}
+
+// WithContext Creates errors that conform to the `HasContext` interface
+type WithContext map[string]interface{}
+
+// Wrapf returns an error annotating err with a stack trace
+// at the point Wrapf is call, and the format specifier.
+// If err is nil, Wrapf returns nil.
+func (c WithContext) Wrapf(err error, format string, args ...interface{}) error {
+	if err == nil {
+		return nil
+	}
+	return &withContext{
+		stack:   callstack.New(1),
+		context: c,
+		cause:   err,
+		msg:     fmt.Sprintf(format, args...),
+	}
+}
+
+// Wrap returns an error annotating err with a stack trace
+// at the point Wrap is called, and the supplied message.
+// If err is nil, Wrap returns nil.
+func (c WithContext) Wrap(err error, msg string) error {
+	if err == nil {
+		return nil
+	}
+	return &withContext{
+		stack:   callstack.New(1),
+		context: c,
+		cause:   err,
+		msg:     msg,
+	}
+}
+
+func (c WithContext) Error(msg string) error {
+	return &withContext{
+		stack:   callstack.New(1),
+		context: c,
+		cause:   errors.New(msg),
+		msg:     "",
+	}
+}
+
+func (c WithContext) Errorf(format string, args ...interface{}) error {
+	return &withContext{
+		stack:   callstack.New(1),
+		context: c,
+		cause:   fmt.Errorf(format, args...),
+		msg:     "",
+	}
+}
+
+// Implements the `error` `causer` and `Contexter` interfaces
+type withContext struct {
+	context WithContext
+	msg     string
+	cause   error
+	stack   *callstack.CallStack
+}
+
+func (c *withContext) Unwrap() error {
+	return c.cause
+}
+
+func (c *withContext) Is(target error) bool {
+	_, ok := target.(*withContext)
```

Comment:
Shouldn't we compare the content too?
```
// Is unwraps its first argument sequentially looking for an error that matches the
// second. It reports whether it finds a match. It should be used in preference to
// simple equality checks:
//
//	if errors.Is(err, fs.ErrExist)
```

Currently, it looks more like `errors.As(err error, target any) bool`.

Reply:
Typical use case is to find a user defined typed error, I expect using `errors.As(err, errors.WithContext{})` will be pretty rare. In addition, if I already know the content of the WithContext{} I'm looking for, why would I need to extract it?